backtracking/coloring.py

```diff
@@ -0,0 +1,118 @@
+from typing import List, Tuple, Union
+
+"""
+    Graph Coloring also called "m coloring problem"
+    consists of coloring given graph with at most m colors
+    such that no adjacent vertices are assigned same color
+
+    Wikipedia: https://en.wikipedia.org/wiki/Graph_coloring
+"""
+
+
+def valid_coloring(
+    neighbours: List[int], colored_vertices: List[int], color: int
+) -> bool:
+    """
+    For each neighbour check if coloring constraint is satisfied
+    If any of the neighbours fail the constraint return False
+    If all neighbours validate constraint return True
+
+    >>> neighbours = [0,1,0,1,0]
+    >>> colored_vertices = [0, 2, 1, 2, 0]
+    
+    >>> color = 1
+    >>> valid_coloring(neighbours, colored_vertices, color)
+    True
+
+    >>> color = 2
+    >>> valid_coloring(neighbours, colored_vertices, color)
+    False
+    """
+    for i in range(len(neighbours)):
+        # Constraint Failed
+        if neighbours[i] == 1 and colored_vertices[i] == color:
+            return False
+    # All neighbours satisfy constraint
+    return True
+
+
+def util_color(
+    graph: List[List[int]], max_colors: int, colored_vertices: List[int], index: int
+) -> bool:
+    """ 
+    Pseudo-Code
+
+    Base Case:
+    1. Check if coloring is complete 
+        1.1 If complete return True (meaning that we successfully colored graph)
+
+    Recursive Step:
+    2. Itterates over each color:
+        Check if current coloring is valid:
+            2.1. Color given vertex
+            2.2. Do recursive call check if this coloring leads to solving problem
+            2.4. if current coloring leads to solution return
+            2.5. Uncolor given vertex
+
+    >>> graph = [[0, 1, 0, 0, 0],
+    ...          [1, 0, 1, 0, 1],
+    ...          [0, 1, 0, 1, 0],
+    ...          [0, 1, 1, 0, 0],
+    ...          [0, 1, 0, 0, 0]]
+    >>> max_colors = 3
+    >>> colored_vertices = [0, 1, 0, 0, 0]
+    >>> index = 3
+    
+    >>> util_color(graph, max_colors, colored_vertices, index)
+    True
+
+    >>> max_colors = 2
+    >>> util_color(graph, max_colors, colored_vertices, index)
+    False
+    """
+
+    # Base Case
+    if index == len(graph):
+        return True
+
+    # Recursive Step
+    for i in range(0, max_colors):
+        if valid_coloring(graph[index], colored_vertices, i):
+            # Color current vertex
+            colored_vertices[index] = i
+            # Validate coloring
+            if util_color(graph, max_colors, colored_vertices, index + 1):
+                return True
+            # Backtrack
+            colored_vertices[index] = -1
+    return False
+
+
+def color(
+    graph: List[List[int]], max_colors: int
+) -> Tuple[bool, Union[List[int], None]]:
```

Comment:
Isn't `typing.Optional[X]` the way to do `Union[X, None]`

Why return `None` when we could just return an empty list?

Reply:
Thanks for responding! 
What do you think is better? Returning empty list of ints or leaving as it is and change to `typing.Optional[X]`

Reply:
Empty list.  We consistently return one data type (a list).

Reply:
Do we need to return two values?  Is there a situation where the return value is True but the list is empty or False but the list is not empty?

Some doctests on `color()` would be a welcome addition.

Reply:
There is only one case, where graph is empty list algorithm return (True, []), but it is edge case and probably no one will use it. We can change it to return one data type and when no coloring is found return empty list. Do you think it's better?

Reply:
I thing it is much easier for the caller if we only return the list.

Reply:
changed :+1: